### ISSUE TITLE
[LSL1] couple of bugs in the club

### ISSUE THREAD

**Issue by vafada** [lanceewing/agile#41]
1. On the dance scene, the floor stays pink instead of multicolor

![Capture](https://user-images.githubusercontent.com/285941/199623022-69e1c747-4831-44d1-8630-19620d451866.PNG)

for reference:

https://youtu.be/Wyvi1pdDPUc?t=845

2. Restoring the game in the club screen doesnt restore all views

on enter room:
![Capture2](https://user-images.githubusercontent.com/285941/199623181-776e671a-418e-4e71-ab23-e8bbdb294a00.PNG)

on restore

![Capture3](https://user-images.githubusercontent.com/285941/199623197-35b85d1d-7b0c-4320-8ece-0c2c7e59519d.PNG)

Save game:

[LLLLLSG.zip](https://github.com/lanceewing/agile/files/9924831/LLLLLSG.zip)

**Comment by vafada** [comment]
The disco floor is not showing up because it has Priority = 3

```
    set.view(o1, View100);
    set.loop(o1, 0);
    set.cel(o1, 0);
    position(o1, 34, 123);
    set.priority(o1, 3);
```

**Comment by vafada** [comment]
@lanceewing the disco floor issue is interesting and will need your advice for this one.

The issue is when Logic 24 loads the disco floor, it sets its priority to 3

```
 set.view(o1, View100);
    set.loop(o1, 0);
    set.cel(o1, 0);
    position(o1, 34, 123);
    set.priority(o1, 3);
```

its being treated as water which make sense if you look at the Picture data for the room

![Capture](https://user-images.githubusercontent.com/285941/199788608-f3b161e3-44a9-452d-9282-2d6dd53580eb.PNG)


and when ego enters it, the dance animation happens.

The disco floor is drawn once Fawn steps on it

```
object.on.water(o5);
    set.view(o5, View99);
    wander(o5);
    draw(o1);
    draw(o2);
    set(f52);
```

Since its priority = 3, AGILE doesn't render it:

https://github.com/lanceewing/agile/blob/75416357961f350c1160c71ce63791e5c649e7cd/AGILE/AnimatedObject.cs#L1465-L1470

because the priority pixel is 4 (AGILE has a separate control pixel array.)


One fix, i did was to do this:

```
// Get the priority colour index for this position from the priority screen.
                        int priorityIndex = state.PriorityPixels[priorityPos];
                        int controlPixel = state.ControlPixels[priorityPos];

                        // If this AnimatedObject's priority is greater or equal to the priority screen value
                        // for this pixel's position, then we'll draw it.
                        // or if the view is in water (control pixel 3)
                        if (this.Priority >= priorityIndex || (this.Priority == 3 && controlPixel == 3))
                        {
```

is this the correct fix?

**Comment by lanceewing** [comment]
It seems that the first pull request has closed this issue automatically.

I'll reopen it while the other PR is being discussed.

**Comment by lanceewing** [comment]
I can't remember if I'd thought about whether the priority of an animated object could be set below 4.

As always, WinAGI has a wealth of info on this, in the documentation for set.priority:

>Valid values for PRI are 3 - 15, with the following considerations:
>
> - If PRI is 15, in addition to having a static priority value of 15, the screen object will also ignore all control lines and blocks. 
> - If PRI is less than 3, the object will not be drawn on the screen, as these values correspond to control lines and are not actually priority values; the object will still exist; it can still move about the screen, and it will cause collisions with other objects;  it will just be invisible. 
> - If PRI is 3, the object will only be visible when on 'water' pixels. 

I wonder if the above is true. It says valid values for PRI are 3 - 15, but then also says that values less than 3 are allowed but the object will be invisible. That implies it is still "valid" but perhaps not useful. Perhaps a programmer might find a use for that; who knows.

I'd like to see if this part of the original AGI interpreter source code exists. Had a quick look but couldn't immediately spot it. 

What I'm wondering is if for PRI values less that 3, that the pixels really are always invisible. What happens if there is a big area of blue priority (i.e. 2) and the object is set to a priority of 2? Is the object really invisible when it passes over the blue area? Or does it behave like priority 3, where pixels on priority 2 are drawn?

Hmmm, I should probably also check if AGILE fully implements the first bullet point above as well. I can see code that ignores control lines when the priority is 15, but haven't spotted anything that ignores blocks when priority is 15. I might need to build a small test game to confirm that the original AGI interpreter and AGILE behave differently in that area.

**Comment by lanceewing** [comment]
> I'd like to see if this part of the original AGI interpreter source code exists. Had a quick look but couldn't immediately spot it.

It looks like IOBJSBRS.ASM might have some code related to this.

**Comment by lanceewing** [comment]
This bit in the WinAGI help makes it clearer:

> There are no priority bands 0 - 2 because these values are used for control lines. Priority 3 (water) is the lowest priority, but is not automatically assigned to screen objects based on movement; priority 4 is the lowest priority automatically assigned based on position, and is given to background elements. All objects not 'on water' will be drawn in front of priority 4 elements regardless of their priority.

So I think that answers all my earlier questions. This is something that, after all these years, I hadn't actually picked up on! i.e. that Priority 3 is actually a "priority". I had always thought of it as one of the control lines, but no, it really is a kind of special priority value.

As the WinAGI help says, it is not automatically assigned to an animated object based on its position. It can only be assigned with set.priority, which is related to the issue you found.

And the original AGI interpreter code does appear to back this one. It is treating only priority 0-2 as being "special", by which it means the control lines:

    IOLoop:
	    mov	al,es:[di]	;get shadow ram (high = priority, low = color)
	    and	al,0F0h		;isolate priority nibble
	    cmp	al,20h		;is this a special priority?
	    jbe	IOSpclPri	;yes

The top 4 bits in the al value are for the priority, so that cmp is checking if it is equal to 2 or below. If so, then jump to the special priority logic. It then searches downwards within the priority screen to find the first pixel below that isn't a special priority:

    IOSpclCK:
	    cmp	di,sbotm_ln	;is current pixel in the bottom line of shadow memory?
	    jae	IOPriCk		;yes - default: shadow priority <= object pri.
    
	    add	di,shdw_lnth	;move down one pixel
	    mov	ch,es:[di]	;get pixel
	    and	ch,0F0h		;isolate shadow priority
	    cmp	ch,20h		;is this one also a special priority
	    jbe	IOSpclCK	;yes

That is the bit that I knew about, but I thought that it also applied to the water priority as well.

So you'll notice something strange in the SplitPriorityPixels method, in the Commands class. This is the bit that is equivalent to the ASM code above. It scans downwards until it finds the first "non-special" priority value...

BUT... there is a "hack" in there supposedly related to KQ1 where if the priority value is 3, then it uses 4 for the priority rather than scan downwards. So I think I must have found a similar issue to what you found, but in KQ1, and handled it as a hack in this method. Previously it would have been scanning down for 3 as well.

That hack was based on a misunderstanding of how priority 3 works with regards to the priority bands. I think this code is wrong:

                    if (data == 3)
                    {
                        state.PriorityPixels[index] = 4;
                        state.ControlPixels[index] = data;
                    }

and perhaps it should instead be:

                    if (data == 3)
                    {
                        state.PriorityPixels[index] = 3;
                        state.ControlPixels[index] = data;
                    }

And likewise this:

                            if (data == 3)
                            {
                                priFound = true;
                                state.PriorityPixels[index] = 4;
                            }

should be:

                            if (data == 3)
                            {
                                priFound = true;
                                state.PriorityPixels[index] = 3;
                            }

I still think it needs its own special "if" in there though, as AGILE needs to put priority 3 into the control pixels, due to the way it works.

So in summary, I suspect the change I proposed above might be another fix for the issue you found. But I agree that the code that you put in the PR would also fix the issue. Given that the code above is wrong though, and probably needs fixing, then it might make your fix redundant.

**Comment by vafada** [comment]
yeah if `state.PriorityPixels[index]` would include value `3` then my PR is redundant since 

`if (this.Priority >= state.PriorityPixels[pixelPos])` will evaluate to `true` for the disco floor

**Comment by lanceewing** [comment]
Yeah, I just tested that change with LSL using your saved game and it works. Given that, I've pushed it to master.

**Comment by vafada** [comment]
both issue fixed
